core/src/main/java/io/grpc/util/CertificateUtils.java

```diff
@@ -0,0 +1,87 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.util;
+
+import com.google.common.base.Charsets;
+import com.google.common.io.BaseEncoding;
+import com.google.common.io.CharStreams;
+import io.grpc.ExperimentalApi;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.security.KeyException;
+import java.security.KeyFactory;
+import java.security.PrivateKey;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.Collection;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Contains certificate/key PEM file utility method(s).
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/8024")
+public final class CertificateUtils {
+  private static CertificateFactory factory;
+  private static final Pattern KEY_PATTERN = Pattern.compile(
+      "-+BEGIN\\s+.*PRIVATE\\s+KEY[^-]*-+(?:\\s|\\r|\\n)+" + // Header
+          "([a-z0-9+/=\\r\\n]+)" +                             // Base64 text
+          "-+END\\s+.*PRIVATE\\s+KEY[^-]*-+",                  // Footer
+      Pattern.CASE_INSENSITIVE);
+  private static final BaseEncoding BASE64_ENCODING = BaseEncoding.base64();
+
+  private static synchronized void initInstance() throws CertificateException {
+    if (factory == null) {
+      factory = CertificateFactory.getInstance("X.509");
+    }
+  }
+
+  /**
+   * Generates X509Certificate array from a file on disk.
+   *
+   * @param inputStream is a {@link InputStream} from the certificate files
+   */
+  public static X509Certificate[] getX509Certificates(InputStream inputStream)
+      throws CertificateException {
+    initInstance();
+    Collection<? extends Certificate> certs = factory.generateCertificates(inputStream);
+    return certs.toArray(new X509Certificate[0]);
+  }
+
+  /** Generates a {@link PrivateKey} from the {@link InputStream}. */
+  public static PrivateKey getPrivateKey(InputStream inputStream)
+      throws Exception {
+    String key = CharStreams.toString(new InputStreamReader(inputStream, Charsets.UTF_8));
+    Matcher m = KEY_PATTERN.matcher(key);
+    if (!m.find()) {
+      throw new KeyException("could not find a PKCS #8 private key in input stream");
+    }
+    // Using System.lineSeparator() here would cause
+    // [Undefined reference | android-api-level-14-4.0_r4] io.grpc.util.(CertificateUtils.java:76)
+    // Is it not supported on Android?
+    String keyContent =
+        m.group(1).replace(System.getProperty("line.separator"), "");
```

Comment:
Is this really platform-dependent? https://tools.ietf.org/html/rfc7468#section-2

> Furthermore, parsers SHOULD ignore whitespace and other non-
   base64 characters and MUST handle different newline conventions.

Later in the section it outlines the characters it means to be whitespace.

Reply:
Yeah, I didn't use `line.separator` before, which passed all tests locally but failed some Windows tests during presubmit.

Reply:
I was saying it should support line endings of both \n and \r\n varieties.

Reply:
Yeah, I think `System.getProperty("line.separator")` should handle both cases: https://stackoverflow.com/questions/36796136/difference-between-system-getpropertyline-separator-and-n

Reply:
No, it doesn't. It only processes Windows newlines on Windows and Unix newlines on Unix. For this it seems we need to support Windows and Unix newlines on Windows and Unix.

Reply:
Oh now I get your idea :) 
I think we could use `line = line.replace("\n", "").replace("\r", "");`, but since we changed to use `readLine()`, I think this parsing is unnecessary here.